File: algo-research/table.py
from seat import Seat

class Table:
    def __init__(self, seats=None):
        self._seats = seats if seats else set()

    def __str__(self):
        value = ""
        for pos, seat in enumerate(self._seats):
            value += "\tSeat #" + str(pos) + "\t|\t" + str(seat) + "\n"

        return value

    def __repr__(self):
        return str(self)

    def __len__(self):
        return len(self._seats)

    def score(self):
        seat_names = self.get_player_names()
        seat_cnt = len(self._seats)

        total_score = 0
        for seat_pos, seat in enumerate(self._seats):
            total_score += seat.score(seat_pos, seat_cnt, seat_names)

        return total_score

    def seat_scores(self):
        seat_names = self.get_player_names()
        seat_cnt = len(self._seats)

        scores = []
        for seat_pos, seat in enumerate(self._seats):
            scores.append((seat, seat.score(seat_pos, seat_cnt, seat_names)))

        return scores

    def meta_score(self, lowest_seat_score):
        seat_names = self.get_player_names()
        seat_cnt = len(self._seats)

        total_score = 0
        for seat_pos, seat in enumerate(self._seats):
            total_score += seat.meta_score(lowest_seat_score, seat_pos, seat_cnt, seat_names)

        return total_score

    def test_seating_scores(self, test_player):
        seat_names = self.get_player_names()
        seat_cnt = len(self._seats)

        candidates = []

        for seat_pos, seat in enumerate(self._seats):
            tmp_player = seat.get_player()

            seat.replace_player(test_player)
            sim_score = seat.score(seat_pos, seat_cnt, seat_names)
            seat.unseat_player()

            if tmp_player:
                seat.seat_player(tmp_player)

            candidates.append((seat, sim_score))

        return candidates

    def get_player_names(self):
        return { s.player_name() for s in self._seats }

    def clone(self):
        return Table([s.clone() for s in self._seats])

    def record_seating(self):
        competitors = self.get_player_names()

        for pos, seat in enumerate(self._seats):
            seat_competitors = competitors.difference({seat.player_name()})
            seat.record_player(pos, seat_competitors)

    def get_unlocked_seats(self):
        return [s for s in self._seats if not s.is_locked()]

    def lock_seats(self):
        for seat in self._seats:
            seat.lock()

    def unlock_seats(self):
        for seat in self._seats:
            seat.unlock()

File: src/app/services/seating/table.js
import _ from 'lodash';

import Seat from './seat';

export default class Table {

  constructor(seats=null) {
    this._seats = seats || []
  }

  to_str() {
    let value = ""

    for(let pos in this._seats) {
      const seat = this._seats[pos];
      value += `\tSeat # ${pos} \t|\t${seat}\n`;
    }

    return value;
  }

  to_repr() {
    return this.to_str();
  }

  count() {
    return this._seats.length;
  }

  get_arrangement() {
    return this._seats.slice(0);
  }

  score() {
    return _.sum(_.map(this.seat_scores(), (s) => s[1]));
  }

  seat_scores() {
    const seated_player_ids = this.get_player_ids()
    const seat_cnt = this.count();

    const scores = [];

    for(let idx in this._seats) {
      const seat = this._seats[idx];
      const seat_pos = parseInt(idx) + 1;

      scores.push([seat, seat.score(seat_pos, seat_cnt, seated_player_ids)]);
    }


    return scores;
  }

  meta_score(lowest_seat_score) {
    const seated_player_ids = this.get_player_ids()
    const seat_cnt = this._seats.length;

    let total_score = 0;

    for(let idx in this._seats) {
      const seat = this._seats[idx];
      const seat_pos = parseInt(idx) + 1;

      total_score += seat.meta_score(lowest_seat_score, seat_pos, seat_cnt, seated_player_ids);
    }

    return total_score;
  }

  test_seating_scores(test_player) {
    const seated_player_ids = this.get_player_ids();
    const seat_cnt = this.count();

    const candidates = [];

    for(let idx in this._seats) {
      const seat = this._seats[idx];
      const seat_pos = parseInt(idx) + 1;

      const tmp_player = seat.get_player();

      let cur_score = 0xFFFFFFFF;
      if(tmp_player !== null) {
        cur_score = seat.score(seat_pos, seat_cnt, seated_player_ids);
      }

      seat.replace_player(test_player);
      const sim_score = seat.score(seat_pos, seat_cnt, seated_player_ids);
      seat.unseat_player();

      if(tmp_player !== null) {
        seat.seat_player(tmp_player);
      }

      candidates.push([seat, sim_score, sim_score <= cur_score])
    }

    return candidates;
  }

  get_player_ids() {
    return _.uniq(_.map(this._seats, (s) => s.player_id()));
  }

  clone() {
    const cloned_seats = _.map(this._seats, (s) => s.clone());
    return new Table(cloned_seats);
  }

  record_seating() {
    const competitors = this.get_player_ids()

    for(let pos in this._seats) {
      const seat = this._seats[pos];
      const seat_competitors = _.difference(competitors, [seat.player_id()]);

      seat.record_player(pos, seat_competitors);
    }
  }

  get_unlocked_seats() {
    return _.filter(this._seats, (s) => !s.is_locked());
  }

  lock_seats() {
    for(let seat of this._seats) {
      seat.lock();
    }
  }

  unlock_seats() {
    for(let seat of this._seats) {
      seat.unlock();
    }
  }
}
